Source/Particles/MultiParticleContainer.cpp

```diff
@@ -461,3 +464,247 @@ MultiParticleContainer::doContinuousInjection() const
+{
+    int i_product;
+    bool found = 0;
+    // Loop over species
+    for (int i=0; i<nspecies; i++){
+        // If species name matches, store its ID
+        // into i_product
+        if (species_names[i] == product_str){
+            found = 1;
+            i_product = i;
+        }
+    }
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE(
+        found != 0,
+        "ERROR: could not find product species ID for ionization. Wrong name?");
+    return i_product;
+}
+
+namespace
+{
+    // For particle i in mfi, if is_ionized[i]=1, copy particle
+    // particle i from container pc_source into pc_product
+    static void createIonizedParticles (
+        int lev, const MFIter& mfi,
+        std::unique_ptr< WarpXParticleContainer>& pc_source,
+        std::unique_ptr< WarpXParticleContainer>& pc_product,
+        amrex::Gpu::ManagedDeviceVector<int>& is_ionized)
+    {
+        BL_PROFILE("createIonizedParticles");
+
+        const int * const AMREX_RESTRICT p_is_ionized = is_ionized.dataPtr();
+
+        const int grid_id = mfi.index();
+        const int tile_id = mfi.LocalTileIndex();
+
+        // Get source particle data
+        auto& ptile_source = pc_source->GetParticles(lev)[std::make_pair(grid_id,tile_id)];
+        const int np_source = ptile_source.GetArrayOfStructs().size();
+        if (np_source == 0) return;
+        // --- source AoS particle data
+        WarpXParticleContainer::ParticleType* particles_source = ptile_source.GetArrayOfStructs()().data();
+        // --- source SoA particle data
+        auto& soa_source = ptile_source.GetStructOfArrays();
+        GpuArray<Real*,PIdx::nattribs> attribs_source;
+        for (int ia = 0; ia < PIdx::nattribs; ++ia) {
+            attribs_source[ia] = soa_source.GetRealData(ia).data();
+        }
+        // --- source runtime attribs
+        GpuArray<Real*,3> runtime_uold_source;
+        // Prepare arrays for boosted frame diagnostics.
+        runtime_uold_source[0] = soa_source.GetRealData(PIdx::ux).data();
+        runtime_uold_source[1] = soa_source.GetRealData(PIdx::uy).data();
+        runtime_uold_source[2] = soa_source.GetRealData(PIdx::uz).data();
+
+        // Indices of product particle for each ionized source particle.
+        // i_product[i]-1 is the location in product tile of product particle
+        // from source particle i.
+        amrex::Gpu::ManagedDeviceVector<int> i_product;
+        i_product.resize(np_source);
+        // 0<i<np_source
+        // 0<i_product<np_ionized
+        // Strictly speaking, i_product should be an exclusive_scan of
+        // is_ionized. However, for indices where is_ionized is 1, the
+        // inclusive scan gives the same result with an offset of 1.
+        // The advantage of inclusive_scan is that the sum of is_ionized
+        // is in the last element, so no other reduction is required to get
+        // number of particles.
+        amrex::Gpu::inclusive_scan(is_ionized.begin(), is_ionized.end(), i_product.begin());
```

Comment:
Could you add a comment here, saying that `Gpu::inclusive_scan` runs on the current GPU stream, and synchronizes with the CPU, so that the next line (executed by the CPU) has the updated values of `i_product`?

Reply:
Comment added.